### PR TITLE
[onert] Align train tensor buffers with EIGEN_MAX_ALIGN_BYTES value

### LINKED ISSUE #11860: [onert] Segfault in sgd optimizer
### What

When I train a model with a softmax layer, a segfault occurs.
This is related to #11658 PR.

<details>
<summary>backtrace</summary>

```
#10 0x00007ffff6436437 in std::_Function_handler<void (long, long), Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_difference_op<float, float>, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer> const, Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const> const, Eigen::ThreadPoolDevice, true, (Eigen::internal::TiledEvaluation)0>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_difference_op<float, float>, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer> const, Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const> const&, Eigen::ThreadPoolDevice const&)::{lambda(long, long)#1}>::_M_invoke(std::_Any_data const&, long&&, std::_Any_data const&) (__functor=..., __args#0=@0x7fffffffc7d0: 0, 
    __args#1=@0x7fffffffc7c8: 15616) at /usr/include/c++/9/bits/std_function.h:300
#11 0x00007ffff6424e1f in std::function<void (long, long)>::operator()(long, long) const (this=0x7fffffffcae0, __args#0=0, __args#1=15616) at /usr/include/c++/9/bits/std_function.h:688
#12 0x00007ffff6422c18 in Eigen::ThreadPoolDevice::parallelFor(long, Eigen::TensorOpCost const&, std::function<long (long)>, std::function<void (long, long)>) const::{lambda(long, long)#1}::operator()(long, long) const (__closure=0x5555556b9c00, firstIdx=0, lastIdx=15616)
    at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h:226
#13 0x00007ffff64290ac in std::_Function_handler<void (long, long), Eigen::ThreadPoolDevice::parallelFor(long, Eigen::TensorOpCost const&, std::function<long (long)>, std::function<void (long, long)>) const::{lambda(long, long)#1}>::_M_invoke(std::_Any_data const&, long&&, std::_Any_data const&) (__functor=..., __args#0=@0x7fffffffc900: 0, __args#1=@0x7fffffffc8f8: 100352)
    at /usr/include/c++/9/bits/std_function.h:300
--Type <RET> for more, q to quit, c to continue without paging--
#14 0x00007ffff6424e1f in std::function<void (long, long)>::operator()(long, long) const (this=0x7fffffffc9a0, __args#0=0, __args#1=100352) at /usr/include/c++/9/bits/std_function.h:688
#15 0x00007ffff6422eda in Eigen::ThreadPoolDevice::parallelFor(long, Eigen::TensorOpCost const&, std::function<long (long)>, std::function<void (long, long)>) const (this=0x55555569cb30, 
    n=100352, cost=..., block_align=..., f=...) at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h:233
#16 0x00007ffff6434fc5 in Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_difference_op<float, float>, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer> const, Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const> const, Eigen::ThreadPoolDevice, true, (Eigen::internal::TiledEvaluation)0>::run (expr=..., device=...) at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorExecutor.h:340
#17 0x00007ffff6434ac8 in Eigen::TensorDevice<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::ThreadPoolDevice>::operator-=<Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> > (
    this=0x7fffffffcc30, other=...) at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorDevice.h:61
#18 0x00007ffff6434706 in nnfw::cker::training_ops::functor::ApplyGradientDescent<Eigen::ThreadPoolDevice, float>::operator()(Eigen::ThreadPoolDevice const&, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::TensorFixedSize<float const, Eigen::Sizes<>, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer>) (this=0x7fffffffccb7, d=..., var=..., lr=..., grad=...) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/training_ops.h:158
#19 0x00007ffff643447f in nnfw::cker::train::GradientDescent (output_shape=..., output_data=0x7ffff6576038, grad_shape=..., grad_data=0x7ffff55ec038, learning_rate=0.00999999978)
    at /home/jyoung/git/npu/ONE/compute/cker/include/cker/train/optimizer/SGD.h:54
#20 0x00007ffff6433a82 in onert::backend::train::optimizer::SGD::applyGradient (this=0x55555569bb00, factors=std::tuple containing = {...})
--Type <RET> for more, q to quit, c to continue without paging--
    at /home/jyoung/git/npu/ONE/runtime/onert/backend/train/optimizer/SGD.cc:53
#21 0x00007ffff64181b5 in onert::backend::train::ops::GradientApplier::applyGradient (this=0x5555556b5920, training_step=0)
    at /home/jyoung/git/npu/ONE/runtime/onert/backend/train/ops/GradientApplier.cc:45
#22 0x00007ffff6ffa703 in onert::exec::train::TrainableFnSequence::backward (this=0x5555556b5c80, training_step=0)
```

</details>

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit aligns train tensor buffers with EIGEN_MAX_ALIGN_BYTES value. Some functions in the eigen library generate the assertion on unaligned arrays.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Fixed #11860

### CODE REVIEW

File: runtime/onert/backend/train/TensorManager.cc [hunk 1]
```diff
@@ -90,7 +90,7 @@ void TensorManager::claimNonConstPlan(const ir::OperandIndex &index)
   auto tensor = _tensors->getNonConstTensor(index);
   assert(tensor && !tensor->is_dynamic());
 
-  auto size = tensor->total_size();
+  auto size = ((tensor->total_size() - 1) | (_align - 1)) + 1;
```
Comment (glistening) [inline runtime/onert/backend/train/TensorManager.cc:93]: @jyoungyun :+1:
Could you please introduce macro like `aligned_size(size, alignment)`?

File: runtime/onert/backend/train/TensorManager.cc [hunk 2]
```diff
@@ -43,6 +43,14 @@ void allocateMemory(backend::train::MemoryManager *mgr,
...
 }
 
+size_t alignedSize(const size_t size, const uint64_t align)
+{
+  assert(size);
+  const size_t aligned = ((size - 1) | (align - 1)) + 1;
+  assert(aligned % 16 == 0);
+  return aligned;
+}
+
```
Comment (glistening) [inline runtime/onert/backend/train/TensorManager.cc:53]: Though it seems size = 0 is not allowed for valid input, we may use the traditional form of:
```cc
#define align_up(num, align) \
    (((num) + ((align) - 1)) & ~((align) - 1))
```
It returns zero when size = 0.

File: runtime/onert/backend/train/TensorManager.h [hunk 3]
```diff
@@ -32,6 +32,13 @@ namespace train
 
 class TensorManager
 {
+public:
+  // Minimum alignment for buffers
+  //
+  // _align is equal to EIGEN_MAX_ALIGN_BYTES, though including Eigen headers
+  // here to get that symbol may not be a good idea.
+  static constexpr uint64_t _align = 16;
```
Comment (ragmani) [inline runtime/onert/backend/train/TensorManager.h:40]: _Inline variables_ are supported since c++17, but onert is still using c++14. So you have to define this variables in the cpp file(`TensorManager.cpp`) to prevent multiple definition error.

File: runtime/onert/backend/train/TensorManager.cc (hunk 2 above)
Comment (ragmani) [reply runtime/onert/backend/train/TensorManager.cc]: Macros are not type safe. So I think it's better to use an inline function although inline functions are not always guaranteed to be inlined.

File: runtime/onert/backend/train/TensorManager.h (hunk 3 above)
Comment (glistening) [reply runtime/onert/backend/train/TensorManager.h:40]: I think it is okay. There are several places in use in our code. `_align` is not odr-used.

File: runtime/onert/backend/train/TensorManager.cc (hunk 2 above)
Comment (glistening) [reply runtime/onert/backend/train/TensorManager.cc]: Ah, I am not saying to use `macro`. Just find the implementation body.

File: runtime/onert/backend/train/TensorManager.h (hunk 3 above)
Comment (jyoungyun) [reply runtime/onert/backend/train/TensorManager.h:40]: @ragmani 

> _Inline variables_ are supported since c++17, but onert is still using c++14. So you have to define this variables in the cpp file(`TensorManager.cpp`) to prevent multiple definition error.

This is a class member variable. Should I need to consider multiple definition error in this case?

Comment (ragmani) [reply runtime/onert/backend/train/TensorManager.h:40]: `static constexpr` means `static constexpr inline` implicitly since c++17.  _inline keyword_ prevents multiple definition by header files that has definitions when linking object files with the header files. So defining `static constexpr` member in a header file is OK in c++17. However, c++14 does not introduce _inline variables_ that can allow using _inline keywords_ to variables such as static member.
If `TensorManager` is generated in multiple object files and the object are linked, c++ compiler will emit multiple definition error.

Comment (glistening) [reply runtime/onert/backend/train/TensorManager.h:40]: As I mentioned in https://github.com/Samsung/ONE/pull/12043#discussion_r1393623601, in this PR's case, it looks okay since there is no odr-use. See https://en.cppreference.com/w/cpp/language/definition#ODR-use. You can find several occurrences the same usage by search "static constexpr" in our code base.

Comment (ragmani) [reply runtime/onert/backend/train/TensorManager.h:40]: I guess there are two ways.
1. Fix it when multiple definition error occors because this code willl not have this problem if onert starts using c++17.
2. Fix it now.

I think No.1 way is not bad.

Comment (glistening) [reply runtime/onert/backend/train/TensorManager.h:40]: @ragmani As I told you offline, `cker/Shape.h` has:

```cc
class Shape
{
  static constexpr int kMaxSmallSize = 5;
```

`cker/Shape.h` is included in almost all `cker/operation/*.h`.
`cker/operation/*.h` is included by the corredponding `*Layer.cc`.

I think the address of `kMaxSmallSize` is not required at all.
Compiler is smart to embed the value w/o preparing memory storage for `kMaxSmallSize`.

Please find 

- my 1st link 

<sub> https://en.cppreference.com/w/cpp/language/definition#ODR-use </sub>
```
ODR-use

Informally,
- an object is odr-used if its value is read (unless it is a compile time constant) or written, its address is taken, or a reference is bound to it, 
```

NOTE that _unless it is a compile time constant_.

- and several SO answers (e.g. https://stackoverflow.com/questions/23491781/when-is-a-variable-odr-used-in-c14)

```
Informally, odr-use of a variable means the following:

If any expression anywhere in the program takes the address of or binds a reference directly to an object, this object must be defined.
```

Comment (ragmani) [reply runtime/onert/backend/train/TensorManager.h:40]: I tried an experiment about it. If `const` or `constexpr` is used, neither "static members" nor "namespace variables" had the problem I mentioned. I don't know the exact reason why `const` or `constexpr` allows "static members" and "namespace variables" to be multi-defined.
Anyway, there is no problem with current code. Sorry for the delay due to incorrect knowledge.

**Review by ragmani — APPROVED** [review]
LGTM
